### PR TITLE
Update QLDB Provider to use Async API

### LINKED ISSUE #388: Update QLDB DataProvider to use the async API
QLDB Driver has been enhanced with [asynchronous APIs](https://github.com/awslabs/amazon-qldb-driver-dotnet/issues/28) so opening this issue to migrate Audit.NET to use it

@thepirat000 I can give it a try in the following days

### PR DESCRIPTION
**Pull request by adrianiftode**
Amazon.QLDB does not support .NET Core 2.0 anymore
unless this dependency is updated to support .NET Core 2.0
https://github.com/amzn/ion-hash-dotnet

Closes #388

### CODE REVIEW

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs [hunk 1]
```diff
@@ -146,23 +147,25 @@ public override Task ReplaceEventAsync(object eventId, AuditEvent auditEvent)
...
                       FROM {tableName} AS e BY eid                      
                       WHERE eid = ?",
                     new ValueFactory().NewString(insertDocumentId));
             });
-
-            var selectedEvent = selectResult.First();
+            var enumerator = selectResult.GetAsyncEnumerator();
+            await enumerator.MoveNextAsync();
+            var selectedEvent = enumerator.Current;
+            await enumerator.DisposeAsync();
```
Comment (thepirat000) [inline src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:168]: We should put the enumerator within an await using block

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs [hunk 2]
```diff
@@ -99,45 +113,32 @@ public override object InsertEvent(AuditEvent auditEvent)
...
         /// <summary>
-        /// Asynchronously inserts an event into AmazonQLDB
+        /// Replaces an event into AmazonQLDB
         /// </summary>
-        public override Task<object> InsertEventAsync(AuditEvent auditEvent)
-        {
-            var result = InsertEvent(auditEvent);
-            return Task.FromResult(result);
-        }
+        public override void ReplaceEvent(object eventId, AuditEvent auditEvent) => ReplaceEventAsync(eventId, auditEvent).GetAwaiter().GetResult();
```
Comment (thepirat000) [inline src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:118]: `.GetResult()` can reasult in deadlock. 
We need a dealock-free version for synchronous context.
We could evaluate providing the option to use both an Async and/or Sync driver?

**Review by thepirat000 — COMMENTED** [review]
Please review comments

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs (hunk 1 above)
Comment (adrianiftode) [reply src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:168]: What this code does is to get the first element and close the stream. With await using we would have to iterate the entire stream.

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs (hunk 2 above)
Comment (adrianiftode) [reply src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:118]: The sync Amazon QLDB implementation also blocks on async code, so we would get into the same issue.

https://github.com/awslabs/amazon-qldb-driver-dotnet/blob/2e3cf8e2d635801732fcc5e5f4b82a9510ccaa38/Amazon.QLDB.Driver/client/Session.cs#L235

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs (hunk 1 above)
Comment (thepirat000) [reply src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:168]: yes, but the difference is that the using statement will call the DisposeAsync even if an exception is thrown right?
Anyway we could simply call `FirstAsync()` method:

`var selectedEvent = await selectResult.FirstAsync();`

File: src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs [hunk 3]
```diff
@@ -74,23 +75,36 @@ public AmazonQldbDataProvider(Action<IAmazonQldbProviderConfigurator> config)
...
+        /// <summary>
+        /// Asynchronously inserts an event into AmazonQLDB
+        /// </summary>
+        public override async Task<object> InsertEventAsync(AuditEvent auditEvent)
         {
             var driver = QldbDriver.Value;
             var tableName = GetTableName(auditEvent);
-
-            List<IIonValue> insertResult = driver.Execute(txn =>
+            var insertResult = new List<IIonValue>();
```
Comment (thepirat000) [inline src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:87]: I think this block could be cleaner if we call `ToListAsync()`:

```c#
var insertResult = await driver.Execute(async txn =>
{
    var json = auditEvent.ToJson();
    var insertInto = $@"INSERT INTO {tableName} VALUE ?";
    try
    {
        return await (await txn.Execute(insertInto, IonLoader.Default.Load(json))).ToListAsync();
    }
    catch (BadRequestException e) when (e.Message.Contains($"No such variable named '{tableName}'"))
    {
        await txn.Execute($"CREATE TABLE {tableName}");
        return await (await txn.Execute(insertInto, IonLoader.Default.Load(json))).ToListAsync();
    }
});
```

Comment (thepirat000) [reply src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs:87]: Or even better, we can call the FirstAsync inside the lambda, and the entire method would be like:

```c#
public override async Task<object> InsertEventAsync(AuditEvent auditEvent)
{
    var driver = QldbDriver.Value;
    var tableName = GetTableName(auditEvent);
    var insertResult = await driver.Execute(async txn =>
    {
        var json = auditEvent.ToJson();
        var insertInto = $@"INSERT INTO {tableName} VALUE ?";
        try
        {
            return await (await txn.Execute(insertInto, IonLoader.Default.Load(json))).FirstAsync();
        }
        catch (BadRequestException e) when (e.Message.Contains($"No such variable named '{tableName}'"))
        {
            await txn.Execute($"CREATE TABLE {tableName}");
            return await (await txn.Execute(insertInto, IonLoader.Default.Load(json))).FirstAsync();
        }
    });

    var insertDocumentId = insertResult.GetField("documentId").StringValue;
    return (insertDocumentId, tableName);
}
```

Comment (adrianiftode) [reply src/Audit.NET.AmazonQLDB/Providers/AmazonQldbDataProvider.cs]: @thepirat000 thanks for checking this; I don't know why, but I had the impression there is no LINQ implementation for async iterators. I made the suggested updates.

**Review by thepirat000 — APPROVED** [review]

**Comment by thepirat000** [comment]
Included on version 17.0.1

**Pull request merged**